internal/controller/instance_controller.go

```diff
@@ -0,0 +1,1628 @@
+package controller
+
+import (
+	"context"
+	_ "embed"
+	"encoding/base64"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"hash/fnv"
+	"path"
+	"strconv"
+	"strings"
+	"time"
+
+	crossplanecommonv1 "github.com/crossplane/crossplane-runtime/apis/common/v1"
+	crossplanemeta "github.com/crossplane/crossplane-runtime/pkg/meta"
+	gcpcloudplatformv1beta1 "github.com/upbound/provider-gcp/apis/cloudplatform/v1beta1"
+	gcpcomputev1beta2 "github.com/upbound/provider-gcp/apis/compute/v1beta2"
+	gcpsecretmanagerv1beta1 "github.com/upbound/provider-gcp/apis/secretmanager/v1beta1"
+	gcpsecretmanagerv1beta2 "github.com/upbound/provider-gcp/apis/secretmanager/v1beta2"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	apimeta "k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	k8sruntime "k8s.io/apimachinery/pkg/runtime"
+	k8sjson "k8s.io/apimachinery/pkg/runtime/serializer/json"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/utils/ptr"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/cluster"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	mcbuilder "sigs.k8s.io/multicluster-runtime/pkg/builder"
+	mccontext "sigs.k8s.io/multicluster-runtime/pkg/context"
+	mcmanager "sigs.k8s.io/multicluster-runtime/pkg/manager"
+	mcreconcile "sigs.k8s.io/multicluster-runtime/pkg/reconcile"
+
+	"go.datum.net/infra-provider-gcp/internal/config"
+	"go.datum.net/infra-provider-gcp/internal/controller/cloudinit"
+	"go.datum.net/infra-provider-gcp/internal/downstreamclient"
+	datumhandler "go.datum.net/infra-provider-gcp/internal/handler"
+	"go.datum.net/infra-provider-gcp/internal/locationutil"
+	datumsource "go.datum.net/infra-provider-gcp/internal/source"
+	networkingv1alpha "go.datum.net/network-services-operator/api/v1alpha"
+	computev1alpha "go.datum.net/workload-operator/api/v1alpha"
+)
+
+const crossplaneFinalizer = "finalizer.managedresource.crossplane.io"
+
+var errResourceIsDeleting = errors.New("resource is deleting")
+
+var imageMap = map[string]string{
+	"datumcloud/ubuntu-2204-lts":           "projects/ubuntu-os-cloud/global/images/ubuntu-2204-jammy-v20240927",
+	"datumcloud/cos-stable-117-18613-0-79": "projects/cos-cloud/global/images/cos-stable-117-18613-0-79",
+}
+
+var machineTypeMap = map[string]string{
+	"datumcloud/d1-standard-2": "n2-standard-2",
+}
```

Comment:
Should we move this to the configuration file now that we can easily support more advanced configuration options easily?